### DISCUSSION TITLE
BACnet MSTP Slave mode doesn't seem to work

### DISCUSSION THREAD

**Discussion by HyperFrank** [bacnet-stack/bacnet-stack#536] [Q&A] [answered]
Hello, 

I have implemented the library in 2 different devices connected on the same RS485 network (both STM32F4). I based my code on the port to this MCU, so my file "dlmstp.c" is the same as the one in "bacnet-stack/ports/stm32f4xx/dlmstp.c". 

My goal is to send a "write property multiple" for a group of 2 to 8 analog outputs from the first device to the second one. When both devices are set as masters, everything works fine. The first device sends the request, the second one receives it, sets its analog outputs to the desired values and sends a simple ACK. 

However, when I try to set the second one as a slave, it captures the request and does nothing with it. Upon request reception, the code reaches the function "MSTP_Slave_Node_FSM()" (instead of "MSTP_Master_Node_FSM()"), it immediately goes to peak "PDU_Queue" that has never been used since boot up and the treatment ends there. No analog output set, no response sent back to the master. 

To setup the slave mode, do I only need to set the mac address over 127, or is there something else I need to setup for this operation to work?

Thank you

**Comment by skarg** [comment]
I think I forgot to add the indication of reception flag - can you add `MSTP_Flag.ReceivePacketPending = true;` to the `MSTP_Slave_Node_FSM()` as shown below to see if that corrects the problem?
```
            case FRAME_TYPE_BACNET_DATA_EXPECTING_REPLY:
                if (DestinationAddress != MSTP_BROADCAST_ADDRESS) {
                    /* The ANSWER_DATA_REQUEST state is entered when a  */
                    /* BACnet Data Expecting Reply, a Test_Request, or  */
                    /* a proprietary frame that expects a reply is received. */
                    pkt = (struct dlmstp_packet *)Ringbuf_Peek(&PDU_Queue);
                    if (pkt != NULL) {
                        MSTP_Send_Frame(pkt->frame_type, pkt->address.mac[0],
                            This_Station, (uint8_t *)&pkt->pdu[0],
                            pkt->pdu_len);
                        Master_State = MSTP_MASTER_STATE_IDLE;
                        /* clear our flag we were holding for comparison */
                        MSTP_Flag.ReceivedValidFrame = false;
                        /* clear the queue */
                        (void)Ringbuf_Pop(&PDU_Queue, NULL);
                    } else if (rs485_silence_elapsed(Treply_delay)) {
                        /* If no reply will be available from the higher layers
                           within Treply_delay after the reception of the final
                           octet of the requesting frame (the mechanism used
                           to determine this is a local matter), then no reply
                           is possible. */
                        MSTP_Flag.ReceivedValidFrame = false;
                    } else {
                        /* indicate successful reception to higher layers */
                        MSTP_Flag.ReceivePacketPending = true;
                    }
                } else {
                    /* no reply when addressed as Broadcast */
                    MSTP_Flag.ReceivedValidFrame = false;
                }
                break;
```

**Reply by HyperFrank** [reply, marked as answer]
Hello, thank you for the reply. 

I copied the dlmstp.c function `MSTP_Slave_Node_FSM()` from the port "bacnet-stack\ports\at91sam7s", which uses the flag you mentioned, `MSTP_Flag.ReceivePacketPending = true;`. Also, in function `dlmstp_receive()`, I had to adapt the clearing of the flag so the `MSTP_Slave_Node_FSM()` could capture the pending state on the next loop: 

    /* if there is a packet that needs processed, do it now. */
    if (MSTP_Flag.ReceivePacketPending) {
        if (This_Station <= 127) {
            MSTP_Flag.ReceivePacketPending = false;
        }
        Statistics.receive_pdu_counter++;
        pdu_len = DataLength;
        src->mac_len = 1;
        src->mac[0] = SourceAddress;
        /* data is already in the pdu pointer */
    }

With that, the operation where the master writes multiple analog values in the slave and receives a simple ACK to continue its own operation works fine. 

However, when in slave mode, using Yabe (Yet Another Bacnet Explorer) on my computer, I see the slave on its MAC address, but it fails to display its BACnet objects... sniffing the UART with a logic analyzer I see an answer that seems correct though. Is it a normal behavior of the BACnet standard, or should I look for another problem?

Thanks again

**Reply by skarg** [reply]
Thank you for the update and the additional change needed in dlmstp_receive() function. I've updated the code and made a pull-request to fix the problems.

It might be normal for YABE - I don't know.  During a BACnet device discovery, an MSTP Slave (aka subordinate nodes) devices do not reply with I-Am to a Who-Is service request or do not reply with an I-Have to a Who-Has service request because the subordinate nodes only respond to Data-Expecting-Reply and Test frames,and those discovery frames are Data-Not-Expecting-Reply frames (typically broadcast).

**Reply by HyperFrank** [reply]
Thanks for the explanation. According to the standard purchased on Ashrae, it indeed seems to be the intended behavior. However, it mentions an optional slave-proxy functionality in which the master provides the I-Am messages on behalf of its salves. Is it supported by the bacnet-stack?

**Reply by skarg** [reply]
The optional MSTP slave-proxy is not supported yet.  MSTP slave proxy is typically implemented by a BACnet/IP (and other datalinks) to MSTP router where there will only be one device on the MSTP segment performing the proxy upstream, but I suppose it could be enabled on any MSTP device on the segment.

**Reply by HyperFrank** [reply]
I see, thank you for the reply and your help!
